src/uct/ib/rc/accel/gga_mlx5.c

```diff
@@ -34,11 +45,157 @@ ucs_config_field_t uct_gga_mlx5_iface_config_table[] = {
   {NULL}
 };
 
+static ucs_status_t
+uct_ib_mlx5_gga_mkey_pack(uct_md_h uct_md, uct_mem_h uct_memh,
+                          void *address, size_t length,
+                          const uct_md_mkey_pack_params_t *params,
+                          void *mkey_buffer)
+{
+    uct_md_mkey_pack_params_t gga_params = *params;
+
+    if (gga_params.field_mask & UCT_MD_MKEY_PACK_FIELD_FLAGS) {
+        gga_params.flags      |= UCT_MD_MKEY_PACK_FLAG_EXPORT;
+    } else {
+        gga_params.field_mask |= UCT_MD_MKEY_PACK_FIELD_FLAGS;
+        gga_params.flags       = UCT_MD_MKEY_PACK_FLAG_EXPORT;
+    }
 
-typedef struct {
-    uct_rc_mlx5_iface_common_t  super;
-} uct_gga_mlx5_iface_t;
+    return uct_ib_mlx5_devx_mkey_pack(uct_md, uct_memh, address, length,
+                                      &gga_params, mkey_buffer);
+}
+
+static ucs_status_t
+uct_gga_mlx5_rkey_unpack(uct_component_t *component, const void *rkey_buffer,
+                         uct_rkey_t *rkey_p, void **handle_p)
+{
+    const uct_ib_md_packed_mkey_t *mkey = rkey_buffer;
+    uct_rkey_bundle_t *rkey_bundle;
+    uct_gga_mlx5_rkey_handle_t *rkey_handle;
+
+    rkey_bundle = ucs_malloc(sizeof(*rkey_bundle), "gga_rkey_bundle");
+    if (rkey_bundle == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    rkey_handle = ucs_malloc(sizeof(*rkey_handle), "gga_rkey_handle");
+    if (rkey_handle == NULL) {
+        ucs_free(rkey_bundle);
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    rkey_handle->packed_mkey    = *mkey;
+    /* memh and rkey_ob will be initialized on demand */
+    rkey_handle->memh           = NULL;
+    rkey_handle->md             = NULL;
+    rkey_handle->rkey_ob.rkey   = UCT_INVALID_RKEY;
+    rkey_handle->rkey_ob.handle = NULL;
+    rkey_handle->rkey_ob.type   = NULL;
+
+    rkey_bundle->handle      = rkey_handle;
+    rkey_bundle->rkey        = (uintptr_t)rkey_bundle;
+    rkey_bundle->type        = NULL;
+
+    *rkey_p   = rkey_bundle->rkey;
+    *handle_p = rkey_bundle->handle;
+    return UCS_OK;
+}
+
+static void
+uct_gga_mlx5_rkey_handle_dereg(uct_gga_mlx5_rkey_handle_t *rkey_handle)
+{
+    uct_md_mem_dereg_params_t params = {
+        .field_mask = UCT_MD_MEM_DEREG_FIELD_MEMH,
+        .memh       = rkey_handle->memh
+    };
+    ucs_status_t status;
+
+    if (rkey_handle->memh == NULL) {
+        return;
+    }
+
+    status = uct_ib_mlx5_devx_mem_dereg(&rkey_handle->md->super.super, &params);
```

Comment:
can we use more internal functions of ib_mlx5? for example, in case of gga there should not be atomic MR

Reply:
Then better to define different memh types as well. WDYT if it will be done in separate PR since it might be large and sensitive change?

Reply:
ok